File: prolog/arithmetic/p06.pl
% P36 (**) Determine the prime factors of a given positive integer (2).

:- ensure_loaded(p05).

factors_mult(X,Y) :-
    X > 0,
    factors_mult(X,Y,2).

factors_mult(1,[],_) :- !.
factors_mult(X,[[Y,M]|Ys],Y) :-
    div(X,Y,M,R), !, % Y divides X.
    next(R,Y,Yn),
    factors_mult(R,Ys,Yn).
factors_mult(X,Y,Z) :-
    !, % Z does not divide X.
    next(X,Z,Zn),
    factors_mult(X,Y,Zn).

div(X,Y,M,R) :-
    div(X,Y,M,R,0),
    M > 0.
div(X,Y,M,R,N) :-
    V is X // Y,
    X =:= V * Y, % Y divides X.
    Nn is N + 1,
    div(V,Y,M,R,Nn).
div(X,_,M,X,M).

File: haskell/arithmetic/p06.hs
-- P36 (**) Determine the prime factors of a given positive integer (2).

import Data.List

primeFactorsMult :: Int -> [(Int,Int)]
primeFactorsMult = map encode . group . primeFactors
  where encode l = (head l, length l)

-- P35
primeFactors :: Int -> [Int]
primeFactors n = factors n 2
  where
    factors 1 _ = []
    factors x y
      | mod x y == 0 = y : factors (div x y) y
      | otherwise    = factors x (y + 1)
